app/controllers/coins_controller.rb

```diff
@@ -56,6 +57,23 @@ def show
 
   protected
 
+  def set_coin
+    coin_id_or_slug = params[:id_or_slug]
+
+    # Attempt to search assuming the param is a slug
+    coin_by_slug = Coin.find_by(slug: coin_id_or_slug)
+    if coin_by_slug
+      @coin = coin_by_slug
+      return
+    end
+
+    # If we don't find matches for slug, we can safely assume it is an id
+    coin_id = coin_id_or_slug
+    coin_by_id = Coin.find(coin_id)
+    # 301 redirect to the same action with the coin slug for SEO purposes
+    redirect_to action: action_name, id_or_slug: coin_by_id.slug, status: :moved_permanently
```

Comment:
Sweet, yeah I wanted to double check with you about that before I made it a permanent redirect

Reply:
Much appreciated, confirmed we want a 301